src/tuple.cr

```diff
@@ -392,6 +392,18 @@ struct Tuple
     pp.list("{", self, "}")
   end
 
+  # Returns a `Array` with all the tuple's elements
+  #
+  # ```
+  # tuple = { 1, 2 }
+  # tuple.to_a # => [1, 2]
+  # ```
+  def to_a
```

Comment:
This can actually be put on `Indexable` to improve all indexable collections instead of just tuple.

Reply:
Doing this means you can remove `Deque#to_a` too, and I'd rather you copied `Deque#to_a`'s implementation using each and preallocating the size than this code which does a bounds-check on every `self[i]`.

Reply:
```
require "benchmark"

struct Tuple
  def to_a1
    Array(Union(*T)).new(size) do |i|
      self[i]
    end
  end

  def to_a2
    arr = Array(Union(*T)).new(size)
    {% for i in 0...T.size %}
      arr << self[{{i}}]
    {% end %}
    arr
  end
end

t = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, "a" }
Benchmark.ips do |x|
  x.report("to_a") { t.to_a }
  x.report("to_a1") { t.to_a1 }
  x.report("to_a2") { t.to_a2 }
end
```

output: 
```
 to_a   1.37M (731.37ns) (± 7.09%)  1.45× slower
to_a1   1.63M (612.81ns) (± 5.71%)  1.21× slower
to_a2   1.98M (505.31ns) (± 8.58%)       fastest
```

Reply:
```
require "benchmark"

module Enumerable
  def to_aa
    ary = Array(T).new(size)
    each { |e| ary << e }
    ary
  end
end

struct Tuple
  def to_a1
    Array(Union(*T)).new(size) do |i|
      self[i]
    end
  end

  def to_a2
    arr = Array(Union(*T)).new(size)
    {% for i in 0...T.size %}
      arr << self[{{i}}]
    {% end %}
    arr
  end
end

t = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, "a" }
Benchmark.ips do |x|
  x.report("to_a") { t.to_a }
  x.report("to_aa") { t.to_aa }
  x.report("to_a1") { t.to_a1 }
  x.report("to_a2") { t.to_a2 }
end
```

output: 
```
 to_a   1.37M (732.25ns) (± 4.60%)  1.51× slower
to_aa   2.04M (489.27ns) (± 9.49%)  1.01× slower
to_a1   1.65M (606.33ns) (± 3.47%)  1.25× slower
to_a2   2.06M ( 485.6ns) (± 8.03%)       fastest
```

Reply:
Conclusion: Skipping optimizing `Tuple#to_a`, will just preallocate the size of the array in `Enumerable#to_a`

Reply:
Please don't use `size` in `Enumerable#to_a`, because `size` traverses the enumerable, so you'll end up traversing it twice.

Let's preallocate in `Indexable` and `Tuple`, but leave `Enumerable#to_a` as is.

Thanks!

Reply:
ok

Reply:
but @asterite don't you think that counting (which only happens if the class doesn't implement their own `size` method) still might be faster than dynamically growing an array?

Reply:
No idea.

Reply:
@carlhoerberg I guess that depends on the implementation of the `Enumerable`. We should assume it's cost intensive by default and use a faster algorithm where `size` trivial (like in `Indexable`).